src/shared/dissect-image.c

```
@@ -395,6 +395,112 @@ static int image_policy_check_protection(
         return 0;
 }
 
+/* internal LUKS2 header defines */
+#define LUKS2_FIXED_HDR_SIZE 0x1000
+#define LUKS2_MAGIC "LUKS\xba\xbe"
+
+/* cryptsetup needs a loop device to work with a partition which has offset/size but
+ * dissect may be running unpriviliged. Implement a minimal custom LUKS header parser
+ * checking integrity protection information. */
+static int partition_is_luks2_integrity(int part_fd, uint64_t offset, uint64_t size) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_free_ char *json = NULL;
+        sd_json_variant *w, *x, *y, *z;
+        const char *key, *type;
+        char luks_magic[sizeof(LUKS2_MAGIC) - 1];
+        be16_t version;
+        be64_t hdr_len;
+        ssize_t sz, json_len;
+        int r;
+
+        if (size < LUKS2_FIXED_HDR_SIZE)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Partition is too small to contain a LUKS header.");
```

should probably return 0 here, since if it's too short to carry a LUKS header it also cannot be integrity protected via luks

---

Previously, we were treating every "crypto_LUKS" partition as "PARTITION_POLICY_ENCRYPTED" without doing any extra checking but since we're adding a parser I decided that we can get some sanity checking for free. I.e. if the partition is small enough to have LUKS header, it is not only treated as "encrypted-without-integrity", but we can actually throw an error and refuse to accept it as encrypted altogether. 

I'm OK with leaving this out for now and returning 0 here but then I believe it should be done in all other corner cases too, e.g. when the declared header length is bogus.